### PR TITLE
KIN-269 Integration of database migrations

### PR DESCRIPTION
**Pull request by CarlosJCD**
[KIN-269](https://loopcrack.atlassian.net/browse/KIN-269) Large [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=loopcrack-org&utm_campaign=badge_info)

### Problem

It was needed to use a way more scalable and maintainable method for creating and modifying the database than an sql script.

### Solution

Implementing codeIgniter 4 migrations for database creation

### Requirements checklist

- [x] It must create all the same tables as the [sql script on the backups directory.](https://github.com/loopcrack-org/Kinub/blob/3210d00badba1353efa01d43d338e8df7eea8332/backup/Database.sql)
- [x] The actual seeders (`UserSeeder`, `CategorySeeder`, `CategoryTagSeeder`, `EmailSeeder`) must work with the generated tables of the migrations file.

#### IMPORTANT:

It is needed to first execute this sql script with your DBMS before using the seeders so that they can work. This is planned to be fixed in a future PR, but right now this script is necessary to test the seeders.
```sql
INSERT INTO `email_types` (`emailTypeId`, `emailTypeName`) VALUES
(1, 'contacto'),
(2, 'soporte tecnico'),
(3, 'info producto');

INSERT INTO `files` (`fileId`, `fileRoute`, `uuid`, `fileName`, `fileDirectoryRoute`) VALUES
(1, '/assets/images/auth-one-bg.jpg', '79c4387d-c1cb-4e6f-b418-8c391d852e2c', 'auth-one-bg.jpg', '/assets/images/'),
(2, '/assets/images/drop.svg', '9d7b76e0-a31f-4e0f-9ac6-b71bc35f2ed4', 'drop.svg', '/assets/images/');

```

#### Testing Steps:

- [x] Validate that the migrations file work by dropping your actual database and on the command line run: 
```sql 
php spark migrate
````
- [x] Validate that the seeders are still working by using each of them:
    - [x] User seeder
    ```sql
    php spark db:seed UserSeeder
    ```

    - [x] Category Seeder
    ```sql
    php spark db:seed CategorySeeder
    ```

    - [x] Category Tag Seeder
    ```sql
    php spark db:seed CategoryTagSeeder
    ```

    - [x] Email Seeder
    ```sql
    php spark db:seed EmailSeeder
    ```

## Resources

- [Database Design Document](https://loopcrack.atlassian.net/wiki/spaces/Kinub/pages/1277953/Database+Design)

## Demo

Screenshot 2023-10-18 at 1 27 58

[KIN-269]: https://loopcrack.atlassian.net/browse/KIN-269?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/Database/Migrations/2023-10-17-220959_Categories.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+use CodeIgniter\Database\Migration;
+
+class Categories extends Migration
+{
+    public function up()
+    {
+        $this->forge->addField([
+            'categoryId' => [
+                'type'           => 'int',
+                'null'           => false,
```
Comment (CheloChidoone) [inline app/Database/Migrations/2023-10-17-220959_Categories.php:14]: @CarlosJCD I don't think this line is necessary, this is the only id of any table where you added this line but it shouldn't be necessary.

**Review by CheloChidoone — CHANGES_REQUESTED** [review]
It looks good @CarlosJCD, I left you a comment about a small error that I think should be fixed

**Review by CheloChidoone — APPROVED** [review]
Great Job @CarlosJCD!

**Review by Cristianpan — APPROVED** [review]

File: app/Database/Migrations/2023-10-17-224407_Users.php [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+                'type'       => 'varchar',
+                'constraint' => 60,
+                'null'       => true,
+            ],
+            'userToken' => [
+                'type'       => 'varchar',
+                'constraint' => 13,
+                'null'       => true,
+            ],
+            'isConfirmed' => [
```
Comment (CheloChidoone) [inline app/Database/Migrations/2023-10-17-224407_Users.php:38]: @CarlosJCD I just noticed something new, now that @FernandoJoachin has added changes to the login logic from the ticket `KIN-276` this field from your migrations breaks the login and other new validations from that ticket; We need to decide if the field is now going to be `isConfirmed` then we need a new ticket for @FernandoJoachin to correct the login logic, the other option is that the name of this field stays just as `confirmed` like before.

CC: @LUGAMAFE @FernandoJoachin

**Review by CheloChidoone — CHANGES_REQUESTED** [review]

File: app/Database/Migrations/2023-10-17-224407_Users.php (hunk 2 above)
Comment (CarlosJCD) [reply app/Database/Migrations/2023-10-17-224407_Users.php:38]: My mistake!, I will change it back to `confirmed` asap

**Review by CheloChidoone — APPROVED** [review]

**Review by CarlosMay7 — APPROVED** [review]

**Review by Mauriciocr207 — COMMENTED** [review]
I see that the migrations do not generate the onDeleteCascase or the Collation to utf8mb4_unicode_c. Is this the expected behavior?

**Comment by CarlosJCD** [comment]
> All good but I see that the migrations do not generate the onDeleteCascase or the Collation to utf8mb4_unicode_c. Is this the expected behavior?

no it is not!, it is supposed to generate a utf8mb4_unicode_c collation and the delete and update options are not supposed to be empty. I will fix it right away, thanks for the comment.

File: app/Database/Seeds/UserSeeder.php [hunk 3]
```diff
@@ -9,29 +10,28 @@ class UserSeeder extends Seeder
...
-                "userPassword" => password_hash("fake_password", PASSWORD_BCRYPT),
-                "userToken" => "FAS12AQajqa",
-                "confirmed" => 0,
-                "isAdmin" => 1
+                'userFirstName' => $faker->name(),
+                'userLastName'  => $faker->lastName(),
+                'userEmail'     => 'admin@admin.com',
+                'userPassword'  => password_hash('fake_password', PASSWORD_BCRYPT),
+                'userToken'     => 'FAS12AQajqa',
+                'isConfirmed'   => 1,
```
Comment (LUGAMAFE) [inline app/Database/Seeds/UserSeeder.php:20]: Based on what you discussed with Marcelo this field now should be called only `confirmed`. Cause this seed is not working. 
```suggestion
                'confirmed'   => 1,
```

File: app/Database/Seeds/UserSeeder.php [hunk 4]
```diff
@@ -9,29 +10,28 @@ class UserSeeder extends Seeder
...
-                "userToken" => "FAS12AQajqa",
-                "confirmed" => 0,
-                "isAdmin" => 0
-            ]
+                'userFirstName' => $faker->name(),
+                'userLastName'  => $faker->lastName(),
+                'userEmail'     => 'nkutch@gmail.com',
+                'userPassword'  => password_hash('fake_password', PASSWORD_BCRYPT),
+                'userToken'     => 'FAS12AQajqa',
+                'isConfirmed'   => 0,
```
Comment (LUGAMAFE) [inline app/Database/Seeds/UserSeeder.php:29]: Same comment as above

**Review by LUGAMAFE — CHANGES_REQUESTED** [review]
Hey @CarlosJCD could you delete the backup .sql file then if we now are going to use the migrations please

**Review by Oswaldo-Chan — APPROVED** [review]

**Review by LUGAMAFE — APPROVED** [review]
Ready @CarlosJCD thanks!

**Pull request merged**